akka-http-core/src/main/scala/akka/http/impl/engine/parsing/HttpRequestParser.scala

```diff
@@ -14,90 +15,122 @@ import akka.http.scaladsl.model._
-              case None ⇒ throw new ParsingException(NotImplemented, ErrorInfo("Unsupported HTTP method", sb.toString))
-            }
-          case c ⇒ parseCustomMethod(ix + 1, sb.append(c))
+  override def createLogic(inheritedAttributes: Attributes): GraphStageLogic = new GraphStageLogic(shape) with HttpMessageParser[RequestOutput] with InHandler with OutHandler {
+
+    import HttpMessageParser._
+
+    override def settings = self.settings
+    override val headerParser = self.headerParser.createShallowCopy()
+
+    private[this] var method: HttpMethod = _
+    private[this] var uri: Uri = _
+    private[this] var uriBytes: Array[Byte] = _
+
+    override def onPush(): Unit = handleParserOutput(parseSessionBytes(grab(in)))
+    override def onPull(): Unit = handleParserOutput(doPull())
+
+    override def onUpstreamFinish(): Unit =
+      if (shouldComplete()) completeStage()
+      else if (isAvailable(out)) handleParserOutput(doPull())
+
+    setHandlers(in, out, this)
+
+    private def handleParserOutput(output: RequestOutput): Unit = {
+      output match {
+        case StreamEnd    ⇒ completeStage()
+        case NeedMoreData ⇒ pull(in)
+        case x            ⇒ push(out, x)
+      }
+    }
+
+    def parseMessage(input: ByteString, offset: Int): StateResult = {
+      var cursor = parseMethod(input, offset)
+      cursor = parseRequestTarget(input, cursor)
+      cursor = parseProtocol(input, cursor)
+      if (byteChar(input, cursor) == '\r' && byteChar(input, cursor + 1) == '\n')
+        parseHeaderLines(input, cursor + 2)
+      else badProtocol
+    }
+
+    def parseMethod(input: ByteString, cursor: Int): Int = {
+      @tailrec def parseCustomMethod(ix: Int = 0, sb: JStringBuilder = new JStringBuilder(16)): Int =
+        if (ix < maxMethodLength) {
+          byteChar(input, cursor + ix) match {
+            case ' ' ⇒
+              customMethods(sb.toString) match {
+                case Some(m) ⇒
+                  method = m
+                  cursor + ix + 1
+                case None ⇒ throw new ParsingException(NotImplemented, ErrorInfo("Unsupported HTTP method", sb.toString))
+              }
+            case c ⇒ parseCustomMethod(ix + 1, sb.append(c))
+          }
+        } else throw new ParsingException(
+          BadRequest,
+          ErrorInfo("Unsupported HTTP method", s"HTTP method too long (started with '${sb.toString}'). " +
+            "Increase `akka.http.server.parsing.max-method-length` to support HTTP methods with more characters."))
+
+      @tailrec def parseMethod(meth: HttpMethod, ix: Int = 1): Int =
+        if (ix == meth.value.length)
+          if (byteChar(input, cursor + ix) == ' ') {
+            method = meth
+            cursor + ix + 1
+          } else parseCustomMethod()
+        else if (byteChar(input, cursor + ix) == meth.value.charAt(ix)) parseMethod(meth, ix + 1)
+        else parseCustomMethod()
+
+      import HttpMethods._
+      byteChar(input, cursor) match {
```

Comment:
btw, make it `(byteChar(input, cursor): @switch) match {` to be sure it's actually a switch? (it just prints a warning, but still)